### ISSUE TITLE
split_non_commuting transform accepts any type of measurement

### ISSUE THREAD

**Issue by Alex-Preciado** [PennyLaneAI/pennylane#4932]
### Important Note

⚠️ This issue is part of an internal assignment and not meant for external contributors.

### Context

Currently, the `split_non_commuting` transform only accepts `expval` and `var` in a circuit. The goal is to remove this limitation and accept any type of measurement: `probs`, `sample` and `counts`. Those measurements accept wires instead of observables.

### Implementation Details

In order to make that work:

1) In [pennylane/transforms/split_non_commuting.py](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/transforms/split_non_commuting.py):

- Update the obs list to take care of cases where there is no observables but wires. Handle `obs.wires` as `PauliZ` observables.
- Update the creation of `new_tape = ` to handle the wires case.
- If necessary update the `reorder` function to support the new measurements

2) Add parametrized tests to ensure the new implementation works with any mix of measurements. [pennylane/tests/transforms/test_split_non_commuting.py](https://github.com/PennyLaneAI/pennylane/blob/master/tests/transforms/test_split_non_commuting.py)

### Requirements

A PR must satisfy the following

- Add unit tests for the transform on tapes
- Add integration tests with QNodes
- Add automatic differentiation tests for all interfaces (Autograd, TF, Torch, Jax and Jax-jit).
- Pass black and pylint checks
- Include a changelog entry under "Improvements"

**Comment by abhishekabhishek** [comment]
Hi @Alex-Preciado,

I am currently working on this issue and just wanted to double check my understanding to make sure I'm on the right path. As I understand currently, the issue is the following:

Consider the following qnode:

```
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(x):
    qml.RX(x,wires=0)
    return qml.probs(wires=[0, 1]), qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliX(0))
    
>>> print(qml.draw(circuit)(0.5))

0: ──RX(0.50)─┤ ╭Probs  <Z>  <X>
1: ───────────┤ ╰Probs
```

Applying `circuit = qml.transforms.split_non_commuting(circuit)` should return a circuit s.t.

```
>>> print(qml.draw(circuit)(0.5))

0: ──RX(0.50)─┤ ╭Probs  <Z>
1: ───────────┤ ╰Probs 
\
0: ──RX(0.50)─┤ <X>
1: ───────────┤
```
But the transform currently doesn't support `SampleMP`, `ProbabilityMP` or `CountsMP` which can take either `wires` or `obs`. I think I have a solution which I'm currently implementing and testing i.e. using `qml.PauliZ(wires[0])@qml.PauliZ(wires[1])@...@qml.PauliZ(wires[len(wires)-1])` as observables when either of these are used only with `wires` argument and I'm looking into the `reorder_fn` to determine if I need to make any changes to it.

Please let me know if you think I missed something :slightly_smiling_face:

Cheers,
Abhi

**Comment by Alex-Preciado** [comment]
Hi @abhishekabhishek , thank you so much for looking into this 🚀 . @rmoyard will be providing some guidance here shortly.

**Comment by rmoyard** [comment]
Hello @abhishekabhishek, thank you for double checking!

Your understanding of the problem is correct, you should remove the limitation concerning those measurements. And yes the main issue is that wires needs to be translated to operators and it seems that you are on a good track!

Let me know if anything is not clear

**Comment by abhishekabhishek** [comment]
Thanks @rmoyard, I am currently writing tests for the modified transform but got a little confused by the existing integration and autodiff tests for e.g. the one below:
https://github.com/PennyLaneAI/pennylane/blob/f435f879113c82edb9a186a69ae86d4b1c137327/tests/transforms/test_split_non_commuting.py#L288-L311

It seems that in this test, the `split_non_commuting` transform does not get called on the Qnode. Am I missing something here? Does the transform automatically gets applied to the Qnode when more than one `expval` are returned?

**Comment by rmoyard** [comment]
Hi @abhishekabhishek, with the old device it is indeed part of the pipeline and called under the hood. With the new device it is not used and split non commuting is not necessary because it can handle non commuting measurements. 

I would suggest to mainly use the transform directly on a tape (qml.tape.QuantumTape) and for integration testing you can apply it directly on the QNode. Even if it is not necessary, the transform will be applied as part of the transform program.

```
     @qml.qnode(dev) 
     def circuit(params): 
         qml.RX(params[0], wires=0) 
         qml.RY(params[1], wires=1) 
         return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)), qml.expval(qml.PauliY(0)))

split_non_commuting(circuit)([0.1, 0.2])
```

In the test that you showed, I would suggest to explicitly apply the transform on the QNodes.

Let me know it that helps!

**Comment by albi3ro** [comment]
Closed by #4972
